tests/actions/SessionTest.js

```diff
@@ -0,0 +1,74 @@
+import {fetchAccountDetails, signIn} from '../../src/libs/actions/Session';
+import * as API from '../../src/libs/API';
+import HttpUtils from '../../src/libs/HttpUtils';
+import waitForPromisesToResolve from '../utils/waitForPromisesToResolve';
+
+// Set up manual mocks for methods used in the actions so our test does not fail.
+jest.mock('../../src/libs/Notification/PushNotification', () => ({
+    register: () => {},
+}));
+
+// We are mocking this method so that we can later test to see if it was called and what arguments it was called with.
+// We test HttpUtils.xhr() since this means that our API command turned into a network request and isn't only queued.
+jest.mock('../../src/libs/HttpUtils', () => ({
+    xhr: jest.fn(),
+}));
+
+test('Authenticate is called with saved credentials when a session expires', () => {
+    const TEST_USER_LOGIN = 'test@testguy.com';
+
+    // Set up mock responses for all APIs that will be called. The next time this command is called it will return
+    // jsonCode: 200 and the response here.
+    API.setMockCommand('GetAccountStatus', jest.fn(() => Promise.resolve({
```

Comment:
I'd love to learn more about the patterns that we're using here. I can't help but notice that there is a different pattern being used for `HttpUtils` and `PushNotification`. What's the difference between the two approaches and why couldn't `jest.mock()` be used for `API` as well?

Reply:
That's a good question. In other words, why can't we do

```js
jest.mock('../src/libs/API', () => ({
    SomeComand: jest.fn(() => Promise.resolve({
        jsonCode: 200,
        accountExists: true,
        canAccessExpensifyCash: true,
        requiresTwoFactorAuth: false,
    }))    
}))
```

- The first problem I ran into when considering this is that API doesn't have a default export. It looks like there is a way around that and I found a SO post that provides some guidance on how to do that [here](https://stackoverflow.com/questions/51203516/mock-named-exports-for-testing-using-jest)
- The second problem is that every API command goes through the `request()` method and that's really the flow we want to test here. i.e. to see what happens when we make a request and it returns `407` and to make sure that what happens after that matches our expectations.

If we replace the entire `API.SomeCommand()` with a mock then the implementation we want to test goes with it.

For the case of `PushNotification` the success or failure of the test doesn't depend on it. We could also just noop there and everything would be OK.

Reply:
If you look at `xhr` however we use the `jest.fn()` so we can make assertions.

Reply:
> The first problem I ran into when considering this is that API doesn't have a default export.

OK, bummer that this only works for default exports. I guess we could always switch API to use `export default` and then change all the `import * as API` to `import API`. If we did that, would it be easier to mock these methods?

I checked out that SO, and I actually like that approach better because it keeps all of the mock logic inside the tests as opposed to putting mock logic inside the lib itself. What are your thoughts on moving that to be inside the tests instead?

I _think_ that the second problem makes sense. There is essentially no good way to mock `request()` unless it was exported from `API`, is that right?

> we use the jest.fn() so we can make assertions.

Ah, OK. I think that was the missing piece I didn't realize. So the idea then is to only use `jest.fn()` when we want to do assertions on a function. I think that's fine for now but it does create some inconsistency in the code (some methods are wrapped, others aren't).

Reply:
>There is essentially no good way to mock request() unless it was exported from API, is that right?

Right so I think we _could_ mock `request()`, but our goal here is to test the logic inside `request()` (which has the queue, retry logic, etc) and the functions that we don't need to run are just `Network.post()` (since we are avoiding calling the API) and `HttpUtils.xhr()` (since we only need to know _if_ we tried to `fetch()` and don't need to actually fetch anything for real).